api/src/main/scala/com/gu/adapters/http/Authentication.scala

```diff
@@ -48,10 +81,69 @@ class AuthenticationService(idapiAuthService: IdapiAuthService) {
 
     // Convert authentication result to return type
     // (identity-auth-core uses cats (Either); discussion-avatar uses scalaz (\/)).
-    result.redeem(
-      err => -\/(userAuthorizationFailed(NonEmptyList(err.getMessage)): Error),
-      identityId => \/-(User(identityId))
-    ).unsafeRunSync()
+    result
+      .redeem(
+        err =>
+          -\/(userAuthorizationFailed(NonEmptyList(err.getMessage)): Error),
+        identityId => \/-(User(identityId))
+      )
+      .unsafeRunSync()
+  }
+
+  private def authenticateUserWithOkta(
+    accessToken: Option[String],
+    identityAccessScope: IdentityAccessScope
+  ): Error \/ User = {
+    // Attempt to authenticate user.
+    val result = for {
+      value <- IO.fromEither(
+        Either.fromOption(
+          accessToken,
+          new Exception("No oauth access token in request")
+        )
+      )
+      credentials = value.stripPrefix("Bearer ")
+      claims <- IO.fromEither(
+        oktaLocalValidator.parsedClaimsFromAccessToken(
+          credentials,
+          List(identityAccessScope),
+          DefaultAccessClaimsParser
+        ).leftMap(e => new Exception(e.message, new Exception(e.suggestedHttpResponseCode.toString)))
+      )
+      identityId = claims.identityId
+    } yield identityId
+
+    // Convert authentication result to return type
+    // (identity-auth-core uses cats (Either); discussion-avatar uses scalaz (\/)).
+    result
+      .redeem(
+        err =>
+          -\/(oauthTokenAuthorizationFailed(NonEmptyList(err.getMessage), err.getCause.getMessage)),
+        identityId => \/-(User(identityId))
+      )
+      .unsafeRunSync()
+  }
+
+  def authenticateUser(
+    scGuUCookie: Option[String],
+    accessToken: Option[String] = None,
+    identityAccessScope: IdentityAccessScope
+  ): Error \/ User = {
+    // check if scGuUCookie or accessToken is present
+    val userOrError = (scGuUCookie, accessToken) match {
+      // if both are present, use okta to authenticate
+      case (Some(_), Some(_)) => authenticateUserWithOkta(accessToken, identityAccessScope)
+      // if only accessToken is present, use okta to authenticate
+      case (None, Some(_)) => authenticateUserWithOkta(accessToken, identityAccessScope)
+      // if only scGuUCookie is present, use idapi to authenticate
+      case (Some(_), None) => authenticateUserWithIdapi(scGuUCookie)
+      // if neither are present, return error
+      case (None, None) =>
+        userAuthorizationFailed(
+          NonEmptyList("No secure cookie or access token in request")
+        ).left
+    }
```

Comment:
I think you'd have one fewer case if you matched on `accessToken` in an outer match and then on `scGuUCookie` in an inner match

Reply:
is this what you're suggesting?

That it doesn't matter what the cookie value is, as long as there is an access token then always use authenticateUserWithOkta?

```suggestion
    val userOrError = (scGuUCookie, accessToken) match {
      // accessToken present regardless of cookie, use okta to authenticate
      case (_, Some(_)) => authenticateUserWithOkta(accessToken, identityAccessScope)
      // if only scGuUCookie is present, use idapi to authenticate
      case (Some(_), None) => authenticateUserWithIdapi(scGuUCookie)
      // if neither are present, return error
      case (None, None) =>
        userAuthorizationFailed(
          NonEmptyList("No secure cookie or access token in request")
        ).left
    }
```

Reply:
Actually I meant something like:
```
accessToken match {
  case Some(_) => 
    ...
  case None =>
    scGuUCookie match {
      ...
}
```
or even, as the matching values aren't being used:
```
val userOrError = if (accessToken.isDefined) 
  authenticateUserWithOkta(accessToken, identityAccessScope)
else if (scGuUCookie.isDefined) 
  authenticateUserWithIdapi(scGuUCookie)
else
 ...
```

Reply:
Ah I prefer those suggestions, I'll refactor the code.